tests/pipelines/test_pipelines_common.py

```diff
@@ -875,84 +875,93 @@ def test_cfg(self):
+            return callback_kwargs
+
+        inputs["callback_on_step_end"] = callback_inputs_change_tensor
+        inputs["callback_on_step_end_tensor_inputs"] = pipe._callback_tensor_inputs
+        inputs["output_type"] = "latent"
         output = pipe(**inputs)[0]
         assert output.abs().sum() == 0
 
     def test_callback_cfg(self):
         sig = inspect.signature(self.pipeline_class.__call__)
+        has_callback_tensor_inputs = "callback_on_step_end_tensor_inputs" in sig.parameters
+        has_callback_step_end = "callback_on_step_end" in sig.parameters
 
-        if not ("callback_on_step_end_tensor_inputs" in sig.parameters and "callback_on_step_end" in sig.parameters):
+        if not (has_callback_tensor_inputs and has_callback_step_end):
             return
 
         if "guidance_scale" not in sig.parameters:
             return
 
         components = self.get_dummy_components()
         pipe = self.pipeline_class(**components)
-        pipe = pipe.to(torch_device)
+        pipe.to(torch_device)
         pipe.set_progress_bar_config(disable=None)
+        self.assertTrue(
+            hasattr(pipe, "_callback_tensor_inputs"),
+            f" {self.pipeline_class} should have `_callback_tensor_inputs` that defines a list of tensor variables its callback function can use as inputs",
+        )
 
-        inputs = self.get_dummy_inputs(torch_device)
-        inputs["guidance_scale"] = 1.0
-        out_no_cfg = pipe(**inputs)[0]
-
-        def callback_no_cfg(pipe, i, t, callback_kwargs):
-            if i == 0:
-                for k, w in callback_kwargs.items():
-                    if k in self.callback_cfg_params:
-                        callback_kwargs[k] = callback_kwargs[k].chunk(2)[-1]
-                pipe._guidance_scale = 1
-            else:
-                # increase guidance scale by one
-                pipe._guidance_scale += 1
+        def callback_increase_guidance(pipe, i, t, callback_kwargs):
+            pipe._guidance_scale += 1.0
 
             return callback_kwargs
 
         inputs = self.get_dummy_inputs(torch_device)
-        inputs["guidance_scale"] = 7.5
-        inputs["callback_on_step_end"] = callback_no_cfg
+
+        # use cfg guidance because some pipelines modify the shape of the latents
+        # outside of the denoising loop
+        inputs["guidance_scale"] = 2.0
+        inputs["callback_on_step_end"] = callback_increase_guidance
         inputs["callback_on_step_end_tensor_inputs"] = pipe._callback_tensor_inputs
-        out_callback_no_cfg = pipe(**inputs)[0]
+        _ = pipe(**inputs)[0]
 
         # we increase the guidance scale by 1.0 at every step
-        assert pipe.guidance_scale == inputs["num_inference_steps"]
-
-        assert out_no_cfg.shape == out_callback_no_cfg.shape
-
-        # outputs won't match since we have different cfg values
-        assert not np.abs(out_no_cfg - out_callback_no_cfg).sum() < 1e-3
```

Comment:
did we remove this test here? I think it is a really good one

Reply:
So the main issue is if we start with no cfg and change it to cfg, a lot of pipelines will break, because some latents are reshaped for cfg guidance outside the denoising loop. 

So changing from no cfg to cfg with the callback, causes a lot of shape mismatch failures. 
https://github.com/huggingface/diffusers/actions/runs/6745445748/job/18337276540?pr=5427#step:11:1058